packages/go_router/lib/src/route.dart

```diff
@@ -279,3 +321,82 @@ class GoRoute {
   ) =>
       const SizedBox.shrink();
 }
+
+/// A route that displays a UI shell around the matching child route. Builds
+/// a new Navigator that is used to display any matching sub-routes, instead
+/// of placing them on the root Navigator.
+///
+/// To display a child route on a different Navigator, provide it with a
+/// [parentNavigatorKey] that matches the [navigatorKey] provided to a
+/// parent [GoRoute] or [ShellRoute] constructor.
+///
+/// The widget built by the matching sub-route becomes the child parameter
+/// of the [builder].
+///
+/// For example:
+///
+/// ```
+/// ShellRoute(
+///   path: '/',
+///   builder: (BuildContext context, GoRouterState state, Widget child) {
+///     return Scaffold(
+///       appBar: AppBar(
+///         title: Text('App Shell')
+///       ),
+///       body: Center(
+///         child: child,
+///       ),
+///     );
+///   },
+///   routes: [
+///     GoRoute(
+///       path: 'a'
+///       builder: (BuildContext context, GoRouterState state) {
+///         return Text('Child Route "/a"');
+///       }
+///     ),
+///   ],
+/// ),
+/// ```
+///
+class ShellRoute extends RouteBase {
+  /// Constructs a [ShellRoute].
+  ShellRoute({
+    required String path,
+    this.builder,
+    this.pageBuilder,
+    this.defaultRoute,
```

Comment:
I started to think the shellRoute should not contribute to path, this will solve a lot of issues, and we won't need defaultRoute. WDYT?

Reply:
What type of issues are you thinking of?

Reply:
We could also make `path` optional in the future.

Reply:
One issue is how to migrate from the navigatorBuilder API. There is also some potential issue when developer enter url that hit the ShellRoute without any subroute. The defaultchild may act like a redirect and update the url. However, user may be stuck when pressing browser backward button.

Reply:
Addressed in https://github.com/flutter/packages/pull/2453/commits/4cdb1834cdd54ee86f3975e800b801b896d109e8